File: main/backend/app/contracts/schemas/writing.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class WritingRequestContext(BaseModel):
    project_key: str | None = Field(default=None, max_length=64)
    trace_id: str | None = Field(default=None, max_length=128)
    request_id: str | None = Field(default=None, max_length=128)
    actor_id: str | None = Field(default=None, max_length=128)
    agent_role: Literal["user_facing_assistant", "orchestration_runtime", "business_capability_wrapper"] | None = Field(
        default=None,
        max_length=64,
    )

    model_config = ConfigDict(extra="forbid")

    @field_validator("project_key")
    @classmethod
    def _project_key_must_not_be_blank(cls, value: str | None) -> str | None:
        normalized = str(value or "").strip()
        return normalized or None


class KeywordCardRequest(WritingRequestContext):
    query: str = Field(..., min_length=1, max_length=200)
    selection_hash: str | None = Field(default=None, max_length=64)
    limit: int = Field(default=10, ge=1, le=50)
    sources: list[Literal["document", "resource", "graph"]] = Field(default_factory=list)
    timeout_ms: int | None = Field(default=None, ge=50, le=30000)
    context: "WritingContextEnvelope | None" = None
    graph_context: dict[str, Any] | None = None

    @field_validator("query")
    @classmethod
    def _query_must_not_be_blank(cls, value: str) -> str:
        normalized = str(value or "").strip()
        if not normalized:
            raise ValueError("query is required")
        return normalized


class KeywordCardItem(BaseModel):
    card_id: str = Field(..., min_length=1, max_length=128)
    source_type: Literal["document", "resource", "graph"]
    title: str = Field(..., min_length=1, max_length=300)
    snippet: str = Field(default="", max_length=2000)
    url: str | None = Field(default=None, max_length=2048)
    score: float = Field(default=0.0, ge=0.0)
    publisher: str | None = Field(default=None, max_length=255)
    published_at: str | None = Field(default=None, max_length=64)
    retrieved_at: str | None = Field(default=None, max_length=64)
    evidence: str | None = Field(default=None, max_length=2000)
    relevance_tags: list[str] = Field(default_factory=list)
    credibility: float | None = Field(default=None, ge=0.0, le=1.0)
    quick_actions: list[str] = Field(default_factory=list)
    extra: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="ignore")


class KeywordCardListResponse(BaseModel):
    cards: list[KeywordCardItem] = Field(default_factory=list)
    selection_hash: str = Field(..., min_length=1, max_length=64)
    suggested_queries: list[str] = Field(default_factory=list)
    search_backends_used: list[str] = Field(default_factory=list)
    source_count: dict[str, int] = Field(default_factory=dict)
    dedupe_count: int = Field(default=0, ge=0)
    score_snapshot: dict[str, Any] = Field(default_factory=dict)
    context_boundary: dict[str, Any] = Field(default_factory=dict)
    dependency_gate: dict[str, Any] = Field(default_factory=dict)
    cache_hit: bool = False
    cache_ttl_ms: int | None = Field(default=None, ge=0)


class KeywordCardPreviewRequest(WritingRequestContext):
    card_id: str = Field(..., min_length=1, max_length=128)
    query: str | None = Field(default=None, max_length=200)


class KeywordCardPreviewResponse(BaseModel):
    card_id: str
    title: str
    url: str | None = None
    publisher: str | None = None
    snippet: str = ""
    score: float = 0.0
    source_type: Literal["document", "resource", "graph"]
    quick_actions: list[str] = Field(default_factory=list)


class KeywordCardDetailRequest(WritingRequestContext):
    card_id: str = Field(..., min_length=1, max_length=128)
    include_provenance: bool = True
    max_provenance_items: int = Field(default=20, ge=1, le=100)


class KeywordCardDetailResponse(BaseModel):
    card_id: str
    title: str
    url: str | None = None
    score: float = 0.0
    evidence: str | None = None
    publisher: str | None = None
    published_at: str | None = None
    retrieved_at: str | None = None
    normalized_query: str | None = None
    dedupe_trace: list[dict[str, Any]] = Field(default_factory=list)
    provenance: dict[str, Any] = Field(default_factory=dict)
    selection_matches: dict[str, Any] = Field(default_factory=dict)
    source_type: Literal["document", "resource", "graph"]


class SuggestRequest(WritingRequestContext):
    query: str = Field(..., min_length=1, max_length=200)
    mode: Literal["keyword", "template", "material", "command"] = "keyword"
    limit: int = Field(default=20, ge=1, le=50)

    @field_validator("query")
    @classmethod
    def _normalize_query(cls, value: str) -> str:
        normalized = str(value or "").strip()
        if not normalized:
            raise ValueError("query is required")
        return normalized


class SuggestItem(BaseModel):
    kind: Literal["keyword", "template", "material", "command"]
    id: str = Field(..., min_length=1, max_length=128)
    label: str = Field(..., min_length=1, max_length=300)
    snippet: str | None = Field(default=None, max_length=500)
    score: float | None = Field(default=None, ge=0.0)
    extra: dict[str, Any] = Field(default_factory=dict)


class SuggestResponse(BaseModel):
    items: list[SuggestItem] = Field(default_factory=list)
    suggest_type: str = Field(..., min_length=1, max_length=64)
    query: str = Field(..., min_length=1, max_length=200)
    source: list[str] = Field(default_factory=list)
    query_rewrite: str = Field(default="", max_length=200)
    selection_hash: str | None = Field(default=None, max_length=64)


class TemplateValidateRequest(WritingRequestContext):
    template_key: str | None = Field(default=None, max_length=128)
    template_content: str | None = None
    template_id: str | None = Field(default=None, max_length=128)
    sample_payload: dict[str, Any] = Field(default_factory=dict)
    strict: bool = False


class TemplateValidateResponse(BaseModel):
    valid: bool
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    normalized_template: dict[str, Any] = Field(default_factory=dict)
    rules: dict[str, Any] = Field(default_factory=dict)
    observability: dict[str, Any] = Field(default_factory=dict)


class LlmActionRequest(WritingRequestContext):
    action_id: Literal["outline_generate", "section_expand", "selection_rewrite", "evidence_summary"]
    template_key: str | None = Field(default=None, max_length=128)
    template_version: str | None = Field(default=None, max_length=64)
    document_id: str | None = Field(default=None, max_length=128)
    input_markdown: str = Field(default="", max_length=50000)
    selection_text: str | None = Field(default=None, max_length=5000)
    target_scope: Literal["selection", "document"] | None = None
    async_mode: bool = Field(default=False, alias="async")
    gate_mode: str | None = Field(default=None, max_length=32)

    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    @model_validator(mode="after")
    def _infer_target_scope(self) -> "LlmActionRequest":
        if self.target_scope is not None:
            return self
        self.target_scope = "selection" if str(self.selection_text or "").strip() else "document"
        return self


class LlmActionResponse(BaseModel):
    content: str = ""
    sources: list[dict[str, Any]] = Field(default_factory=list)
    mode: str = ""
    warnings: list[str] = Field(default_factory=list)
    trace_id: str | None = None
    job_id: int | None = None
    status: str = "completed"
    capability_truth: dict[str, Any] = Field(default_factory=dict)
    observability: dict[str, Any] = Field(default_factory=dict)
    action_boundary: dict[str, Any] = Field(default_factory=dict)
    dependency_gate: dict[str, Any] = Field(default_factory=dict)


class LlmActionHistoryItem(BaseModel):
    job_id: int
    job_type: str
    status: str
    project_key: str | None = None
    action_id: str | None = None
    template_key: str | None = None
    template_version: str | None = None
    request_meta: dict[str, Any] = Field(default_factory=dict)
    actor_id: str | None = None
    trace_id: str | None = None
    created_at: str | None = None
    duration_ms: int | None = None
    result_summary: dict[str, Any] = Field(default_factory=dict)


class WritingDocumentData(BaseModel):
    id: int | None = None
    project_key: str | None = None
    title: str = ""
    body_md: str = ""
    status: str | None = "draft"
    version: int | None = None
    etag: str | None = None
    updated_by_user_id: str | None = None
    updated_at: str | None = None
    created_at: str | None = None
    metadata_json: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="allow")


class WritingDocumentListData(BaseModel):
    items: list[WritingDocumentData] = Field(default_factory=list)

    model_config = ConfigDict(extra="allow")


class WritingDocumentDeleteData(BaseModel):
    deleted: bool = False
    document: WritingDocumentData | dict[str, Any] = Field(default_factory=WritingDocumentData)

    model_config = ConfigDict(extra="allow")


class WritingDocumentDraftData(BaseModel):
    id: int | None = None
    doc_id: int | None = None
    project_key: str | None = None
    draft_body_md: str = ""
    selection_snapshot: dict[str, Any] | list[Any] | None = Field(default_factory=dict)
    base_version: int | None = None
    autosave_token: str | None = None
    request_id: str | None = None
    updated_at: str | None = None
    created_at: str | None = None

    model_config = ConfigDict(extra="allow")


class WritingCitationData(BaseModel):
    id: int | None = None
    doc_id: int | None = None
    project_key: str | None = None
    source_doc_id: int | None = None
    source_uri: str | None = None
    source_title: str | None = None
    quote_text: str | None = None
    position_anchor: str | None = None
    card_id: str | None = None
    metadata_json: dict[str, Any] | None = Field(default_factory=dict)
    created_at: str | None = None
    updated_at: str | None = None

    model_config = ConfigDict(extra="allow")


class WritingCitationListData(BaseModel):
    items: list[WritingCitationData] = Field(default_factory=list)

    model_config = ConfigDict(extra="allow")


PrimaryWritingLoopStage = Literal[
    "document_ready",
    "editing",
    "saved",
    "context_loaded",
    "citation_applied",
    "action_executed",
    "write_back_ready",
]


class WritingBaselineCapability(BaseModel):
    capability_id: str = Field(..., min_length=1, max_length=64)
    designed: bool
    implemented: bool
    still_open: bool
    owner_modules: list[str] = Field(default_factory=list)
    notes: str = Field(default="", max_length=500)

    model_config = ConfigDict(extra="forbid")


class WritingBaselineDeltaMatrix(BaseModel):
    contract_version: str = Field(default="writing.wave_a.e1.v1", min_length=1, max_length=64)
    capabilities: list[WritingBaselineCapability] = Field(default_factory=list)
    non_goals: list[str] = Field(default_factory=list)
    repo_reality: str = Field(default="", max_length=500)

    model_config = ConfigDict(extra="forbid")


class WritingPrimaryLoopCheckpoint(BaseModel):
    project_key: str | None = Field(default=None, max_length=64)
    document_id: int | None = Field(default=None, ge=1)
    has_markdown_body: bool = False
    saved_version: int | None = Field(default=None, ge=1)
    has_context_cards: bool = False
    has_accepted_citation: bool = False
    llm_action_invoked: bool = False
    has_write_back_candidate: bool = False
    graph_context_attached: bool = False
    graph_handoff_contract_version: str | None = Field(default=None, max_length=64)
    llm_consumer: str | None = Field(default="writing.llm_action", max_length=128)
    frontend_surface: str | None = Field(default="writing.workbench", max_length=128)

    model_config = ConfigDict(extra="forbid")


class WritingPrimaryLoopState(BaseModel):
    contract_version: str = Field(default="writing.wave_a.e2.v1", min_length=1, max_length=64)
    stages: list[PrimaryWritingLoopStage] = Field(default_factory=list)
    next_required_stage: PrimaryWritingLoopStage | None = None
    always_on_layers: list[str] = Field(default_factory=list)
    optional_layers: list[str] = Field(default_factory=list)
    no_graph_happy_path_complete: bool = False
    selection_level_entry_supported: bool = True
    document_level_entry_supported: bool = True
    ordering_violations: list[str] = Field(default_factory=list)
    cross_theme_dependency_gate: dict[str, Any] = Field(default_factory=dict)

    model_config = ConfigDict(extra="forbid")


class WritingContextEnvelope(BaseModel):
    contract_version: str = Field(default="writing.context_boundary.e3.v1", min_length=1, max_length=64)
    selection_context: dict[str, Any] = Field(default_factory=dict)
    evidence_context: dict[str, Any] = Field(default_factory=dict)
    accepted_citation_context: dict[str, Any] = Field(default_factory=dict)
    graph_context: dict[str, Any] | None = None

    model_config = ConfigDict(extra="forbid")

File: main/frontend-modern/src/lib/api/domains/writing.ts
import { endpoints } from '../endpoints'
import { apiClient, asList, httpDelete as del, httpGet as get, httpPatch as patch, httpPost as post, unwrapEnvelope } from '../client'

export type WritingDocumentStatus = 'draft' | 'published' | 'archived' | string

export type WritingDocument = {
  id: number
  project_key: string
  title: string
  body_md: string
  status: WritingDocumentStatus
  version: number
  etag: string
  updated_by_user_id?: string | null
  updated_at?: string | null
  created_at?: string | null
  metadata_json: Record<string, unknown>
}

export type WritingDraft = {
  id: number
  doc_id: number
  project_key: string
  draft_body_md: string
  selection_snapshot?: Record<string, unknown> | null
  base_version: number
  autosave_token: string
  request_id?: string | null
  updated_at?: string | null
  created_at?: string | null
}

export type WritingCitation = {
  id?: number
  doc_id?: number
  project_key?: string
  source_doc_id?: number | null
  source_uri?: string | null
  source_title?: string | null
  quote_text?: string | null
  position_anchor?: string | null
  card_id?: string | null
  metadata_json?: Record<string, unknown> | null
  created_at?: string | null
  updated_at?: string | null
}

export type WritingTemplate = {
  template_key: string
  label: string
  description?: string | null
  template_content: string
}

export type WritingTemplateValidation = {
  valid: boolean
  errors: string[]
  warnings: string[]
  normalized_template: Record<string, unknown>
  rules: Record<string, unknown>
  observability: Record<string, unknown>
}

export type TypedKnowledgeWritingHandoff = {
  contract_version: 'typed_knowledge.writing_handoff.v1'
  knowledge_item_key: string
  project_key: string
  canonical_statement: string
  primary_type_node_key: string
  topic_cluster_keys: string[]
  booklet_keys: string[]
  review_state: string
  quality_grade?: string | null
  locale?: string | null
  evidence_refs: string[]
  visibility_scope: string
  selection_hash?: string | null
  selection_text?: string | null
  facets: Record<string, unknown>
}

export type TypedKnowledgeWritingContext = {
  contract_version: 'writing.typed_knowledge_context.v1'
  source: 'typed_knowledge'
  consumer: 'writing.keyword_card'
  handoffs: TypedKnowledgeWritingHandoff[]
  boundary: Record<string, unknown>
}

export type WritingContextEnvelope = {
  contract_version?: string
  selection_context?: Record<string, unknown>
  evidence_context?: Record<string, unknown>
  accepted_citation_context?: Record<string, unknown>
  graph_context?: Record<string, unknown> | null
  typed_knowledge_context?: TypedKnowledgeWritingContext | null
}

export type WritingKeywordCardSource = 'document' | 'resource' | 'graph'

export type WritingKeywordCard = {
  card_id: string
  source_type: WritingKeywordCardSource
  title: string
  snippet: string
  url?: string | null
  score: number
  publisher?: string | null
  published_at?: string | null
  retrieved_at?: string | null
  evidence?: string | null
  relevance_tags: string[]
  credibility?: number | null
  quick_actions: string[]
  extra: Record<string, unknown>
}

export type WritingKeywordCardListResponse = {
  cards: WritingKeywordCard[]
  selection_hash: string
  suggested_queries: string[]
  search_backends_used: string[]
  source_count: Record<string, number>
  dedupe_count: number
  score_snapshot: Record<string, unknown>
  context_boundary: Record<string, unknown>
  dependency_gate: Record<string, unknown>
  cache_hit: boolean
  cache_ttl_ms?: number | null
}

export type WritingKeywordCardPreview = {
  card_id: string
  title: string
  url?: string | null
  publisher?: string | null
  snippet: string
  score: number
  source_type: WritingKeywordCardSource
  quick_actions: string[]
}

export type WritingKeywordCardDetail = {
  card_id: string
  title: string
  url?: string | null
  score: number
  evidence?: string | null
  publisher?: string | null
  published_at?: string | null
  retrieved_at?: string | null
  normalized_query?: string | null
  dedupe_trace: Array<Record<string, unknown>>
  provenance: Record<string, unknown>
  selection_matches: Record<string, unknown>
  source_type: WritingKeywordCardSource
}

export type WritingSuggestMode = 'keyword' | 'template' | 'material' | 'command'

export type WritingSuggestItem = {
  kind: WritingSuggestMode
  id: string
  label: string
  snippet?: string | null
  score?: number | null
  extra: Record<string, unknown>
}

export type WritingSuggestResponse = {
  items: WritingSuggestItem[]
  suggest_type: string
  query: string
  source: string[]
  query_rewrite: string
  selection_hash?: string | null
}

export type WritingLlmActionId = 'outline_generate' | 'section_expand' | 'selection_rewrite' | 'evidence_summary'

export type WritingLlmActionResponse = {
  content: string
  sources: Array<Record<string, unknown>>
  mode: string
  warnings: string[]
  trace_id?: string | null
  job_id?: number | null
  status: string
  capability_truth: Record<string, unknown>
  observability: Record<string, unknown>
  action_boundary: Record<string, unknown>
  dependency_gate: Record<string, unknown>
}

export type WritingLlmActionHistoryItem = {
  job_id: number
  job_type: string
  status: string
  project_key?: string | null
  action_id?: string | null
  template_key?: string | null
  template_version?: string | null
  request_meta: Record<string, unknown>
  actor_id?: string | null
  trace_id?: string | null
  created_at?: string | null
  duration_ms?: number | null
  result_summary: Record<string, unknown>
}

export type CreateWritingDocumentPayload = {
  project_key?: string
  title: string
  body_md?: string
  updated_by_user_id?: string | null
  metadata_json?: Record<string, unknown>
}

export type UpdateWritingDocumentPayload = {
  project_key?: string
  title?: string
  body_md: string
  base_version?: number | null
  updated_by_user_id?: string | null
  metadata_json?: Record<string, unknown> | null
}

export type AutosaveWritingDraftPayload = {
  project_key?: string
  draft_body_md: string
  base_version?: number | null
  autosave_token: string
  request_id?: string | null
  selection_snapshot?: Record<string, unknown>
}

export type ValidateWritingTemplatePayload = {
  project_key?: string
  template_key?: string | null
  template_content?: string | null
  template_id?: string | null
  sample_payload?: Record<string, unknown>
  strict?: boolean
}

export type WritingKeywordCardRequest = {
  project_key?: string
  trace_id?: string | null
  request_id?: string | null
  query: string
  selection_hash?: string | null
  limit?: number
  sources?: WritingKeywordCardSource[]
  timeout_ms?: number | null
  context?: WritingContextEnvelope | null
  graph_context?: Record<string, unknown> | null
}

export type WritingKeywordCardPreviewRequest = {
  project_key?: string
  trace_id?: string | null
  request_id?: string | null
  card_id: string
  query?: string | null
}

export type WritingCardDetailParams = {
  project_key?: string
  include_provenance?: boolean
  max_provenance_items?: number
}

export type WritingSuggestParams = {
  mode?: WritingSuggestMode
  project_key?: string
  limit?: number
}

export type WritingLlmActionPayload = {
  project_key?: string
  trace_id?: string | null
  request_id?: string | null
  action_id: WritingLlmActionId
  template_key?: string | null
  template_version?: string | null
  document_id?: string | null
  input_markdown?: string
  selection_text?: string | null
  target_scope?: 'selection' | 'document' | null
  async?: boolean
  gate_mode?: string | null
}

export const WRITING_CONTEXT_BOUNDARY_CONTRACT_VERSION = 'writing.context_boundary.e3.v1' as const
export const WRITING_TYPED_KNOWLEDGE_CONTEXT_VERSION = 'writing.typed_knowledge_context.v1' as const
export const WRITING_TYPED_KNOWLEDGE_HANDOFF_CONTRACT_VERSION = 'typed_knowledge.writing_handoff.v1' as const
export const WRITING_TYPED_KNOWLEDGE_CONSUMER = 'writing.keyword_card' as const

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value))
}

function stringValue(value: unknown) {
  return typeof value === 'string' ? value.trim() : ''
}

function stringList(value: unknown) {
  if (!Array.isArray(value)) return []
  return value.map((item) => stringValue(item)).filter(Boolean)
}

function typedKnowledgeContextCandidate(value: unknown): unknown {
  if (!isRecord(value)) return null
  if (value.contract_version === WRITING_TYPED_KNOWLEDGE_CONTEXT_VERSION) return value
  for (const key of ['typed_knowledge_context', 'writing_typed_knowledge_context']) {
    if (key in value) return value[key]
  }
  for (const key of ['writing_context', 'context', 'metadata_json']) {
    const nested = value[key]
    if (isRecord(nested)) {
      const candidate = typedKnowledgeContextCandidate(nested)
      if (candidate) return candidate
    }
  }
  return null
}

function normalizeTypedKnowledgeWritingHandoff(value: unknown): TypedKnowledgeWritingHandoff | null {
  if (!isRecord(value) || value.contract_version !== WRITING_TYPED_KNOWLEDGE_HANDOFF_CONTRACT_VERSION) return null
  const knowledge_item_key = stringValue(value.knowledge_item_key)
  const project_key = stringValue(value.project_key)
  const canonical_statement = stringValue(value.canonical_statement)
  const primary_type_node_key = stringValue(value.primary_type_node_key)
  const evidence_refs = stringList(value.evidence_refs)
  const review_state = stringValue(value.review_state)
  const visibility_scope = stringValue(value.visibility_scope)
  if (!knowledge_item_key || !project_key || !canonical_statement || !primary_type_node_key || !review_state || !visibility_scope || !evidence_refs.length) {
    return null
  }
  return {
    contract_version: WRITING_TYPED_KNOWLEDGE_HANDOFF_CONTRACT_VERSION,
    knowledge_item_key,
    project_key,
    canonical_statement,
    primary_type_node_key,
    topic_cluster_keys: stringList(value.topic_cluster_keys),
    booklet_keys: stringList(value.booklet_keys),
    review_state,
    quality_grade: stringValue(value.quality_grade) || null,
    locale: stringValue(value.locale) || null,
    evidence_refs,
    visibility_scope,
    selection_hash: stringValue(value.selection_hash) || null,
    selection_text: stringValue(value.selection_text) || null,
    facets: isRecord(value.facets) ? value.facets : {},
  }
}

export function readTypedKnowledgeWritingContext(value: unknown): TypedKnowledgeWritingContext | null {
  const candidate = typedKnowledgeContextCandidate(value)
  if (!isRecord(candidate)) return null
  if (candidate.contract_version !== WRITING_TYPED_KNOWLEDGE_CONTEXT_VERSION) return null
  if (candidate.source !== 'typed_knowledge' || candidate.consumer !== WRITING_TYPED_KNOWLEDGE_CONSUMER) return null
  const handoffs = Array.isArray(candidate.handoffs)
    ? candidate.handoffs.map((item) => normalizeTypedKnowledgeWritingHandoff(item)).filter((item): item is TypedKnowledgeWritingHandoff => Boolean(item))
    : []
  if (!handoffs.length) return null
  return {
    contract_version: WRITING_TYPED_KNOWLEDGE_CONTEXT_VERSION,
    source: 'typed_knowledge',
    consumer: WRITING_TYPED_KNOWLEDGE_CONSUMER,
    handoffs,
    boundary: isRecord(candidate.boundary) ? candidate.boundary : {},
  }
}

export function readTypedKnowledgeWritingContextFromDocument(document: Pick<WritingDocument, 'metadata_json'> | null | undefined) {
  return readTypedKnowledgeWritingContext(document?.metadata_json)
}

export function writingTypedKnowledgeContextKey(context: TypedKnowledgeWritingContext | null | undefined) {
  if (!context) return 'typed_knowledge:none'
  const handoffKeys = context.handoffs.map((handoff) => [
    handoff.knowledge_item_key,
    handoff.selection_hash || '',
    handoff.visibility_scope,
  ].join(':'))
  return `typed_knowledge:${handoffKeys.join('|')}`
}

export function withTypedKnowledgeWritingContext(
  payload: WritingKeywordCardRequest,
  context: TypedKnowledgeWritingContext | null | undefined,
): WritingKeywordCardRequest {
  if (!context) return payload
  const existingContext: WritingContextEnvelope = isRecord(payload.context) ? (payload.context as WritingContextEnvelope) : {}
  return {
    ...payload,
    context: {
      ...existingContext,
      contract_version: existingContext.contract_version || WRITING_CONTEXT_BOUNDARY_CONTRACT_VERSION,
      typed_knowledge_context: context,
    },
  }
}

export function buildPersistedTypedKnowledgeKeywordCardRequest({
  projectKey,
  query,
  selectionHash,
  document,
  limit,
  sources = ['document', 'resource', 'graph'],
}: {
  projectKey: string
  query: string
  selectionHash?: string | null
  document: Pick<WritingDocument, 'metadata_json'> | null | undefined
  limit?: number
  sources?: WritingKeywordCardSource[]
}): WritingKeywordCardRequest {
  const typedContext = readTypedKnowledgeWritingContextFromDocument(document)
  return withTypedKnowledgeWritingContext(
    {
      project_key: projectKey,
      query,
      selection_hash: selectionHash || undefined,
      limit,
      sources,
    },
    typedContext,
  )
}

function withQuery(path: string, params: Record<string, string | number | boolean | null | undefined>) {
  const query = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value == null || value === '') return
    query.set(key, String(value))
  })
  const suffix = query.toString()
  return suffix ? `${path}?${suffix}` : path
}

function listResponse<T>(value: T[] | { items?: T[] }) {
  return asList<T>(value)
}

export async function listWritingDocuments(limit = 50) {
  const data = await get<WritingDocument[] | { items?: WritingDocument[] }>(withQuery(endpoints.writing.documents, { limit }))
  return listResponse<WritingDocument>(data)
}

export async function createWritingDocument(payload: CreateWritingDocumentPayload) {
  return post<WritingDocument>(endpoints.writing.documents, {
    body_md: '',
    metadata_json: {},
    ...payload,
  })
}

export async function getWritingDocument(docId: number) {
  return get<WritingDocument>(endpoints.writing.documentById(docId))
}

export async function deleteWritingDocument(docId: number, projectKey?: string) {
  return del<{ deleted: boolean; document: WritingDocument }>(
    withQuery(endpoints.writing.documentById(docId), { project_key: projectKey }),
  )
}

export async function updateWritingDocument(
  docId: number,
  payload: UpdateWritingDocumentPayload,
  options?: { ifMatch?: string | null },
) {
  if (!options?.ifMatch) {
    return patch<WritingDocument>(endpoints.writing.documentById(docId), payload)
  }

  const response = await apiClient.patch(endpoints.writing.documentById(docId), payload, {
    headers: { 'If-Match': options.ifMatch },
  })
  return unwrapEnvelope<WritingDocument>(response.data)
}

export async function autosaveWritingDraft(docId: number, payload: AutosaveWritingDraftPayload) {
  return post<WritingDraft>(endpoints.writing.documentDraft(docId), payload)
}

export async function listWritingCitations(docId: number) {
  const data = await get<WritingCitation[] | { items?: WritingCitation[] }>(endpoints.writing.documentCitations(docId))
  return listResponse<WritingCitation>(data)
}

export async function upsertWritingCitations(docId: number, citations: WritingCitation[], projectKey?: string) {
  const data = await post<{ items?: WritingCitation[] }>(endpoints.writing.documentCitations(docId), {
    project_key: projectKey,
    citations,
  })
  return listResponse<WritingCitation>(data)
}

export async function listWritingTemplates() {
  const data = await get<WritingTemplate[] | { items?: WritingTemplate[] }>(endpoints.writing.templates)
  return listResponse<WritingTemplate>(data)
}

export async function validateWritingTemplate(payload: ValidateWritingTemplatePayload) {
  return post<WritingTemplateValidation>(endpoints.writing.templateValidate, {
    sample_payload: {},
    strict: false,
    ...payload,
  })
}

export async function getWritingKeywordCards(payload: WritingKeywordCardRequest) {
  return post<WritingKeywordCardListResponse>(endpoints.writing.keywordCards, payload)
}

export async function previewWritingKeywordCard(payload: WritingKeywordCardPreviewRequest) {
  return post<WritingKeywordCardPreview>(endpoints.writing.keywordCardPreview, payload)
}

export async function getWritingCardDetail(cardId: string, params: WritingCardDetailParams = {}) {
  return get<WritingKeywordCardDetail>(
    withQuery(endpoints.writing.cardById(cardId), {
      project_key: params.project_key,
      include_provenance: params.include_provenance ?? true,
      max_provenance_items: params.max_provenance_items ?? 20,
    }),
  )
}

export async function getWritingSuggest(query: string, params: WritingSuggestParams = {}) {
  return get<WritingSuggestResponse>(
    withQuery(endpoints.writing.suggest, {
      query,
      mode: params.mode ?? 'keyword',
      project_key: params.project_key,
      limit: params.limit ?? 20,
    }),
  )
}

export async function runWritingLlmAction(payload: WritingLlmActionPayload) {
  return post<WritingLlmActionResponse>(endpoints.writing.llmActions, payload)
}

export async function listWritingLlmActionHistory(limit = 20) {
  const data = await get<WritingLlmActionHistoryItem[] | { items?: WritingLlmActionHistoryItem[] }>(
    withQuery(endpoints.writing.llmActionHistory, { limit }),
  )
  return listResponse<WritingLlmActionHistoryItem>(data)
}

export async function getWritingLlmActionDetail(jobId: number) {
  return get<WritingLlmActionHistoryItem>(endpoints.writing.llmActionById(jobId))
}

export async function exportWritingMarkdown(docId: number, projectKey?: string) {
  const response = await apiClient.post<string>(
    endpoints.writing.exportMarkdown,
    {
      doc_id: docId,
      project_key: projectKey,
    },
    {
      responseType: 'text',
    },
  )
  const contentDisposition = String(response.headers['content-disposition'] || '')
  const match = contentDisposition.match(/filename=([^;]+)/i)
  return {
    filename: match?.[1]?.replace(/^"+|"+$/g, '') || `writing-document-${docId}.md`,
    markdown: typeof response.data === 'string' ? response.data : '',
  }
}
